### PR TITLE
Issue 1339 - API Key support

### LINKED ISSUE #1339: Add api key support
### Description
Add api key support for enabling the use of the api from a third party application without the need of  being logged in.

### Goals
- [x] Users can create an api key for personal use
- [ ] Users can regenerate or remove an api key
- [x] Appending the api key to a request will work as if the user is logged in

### Tasks
- [x] Add API key support in the backend for impersonate a user by appending the api-key as a query param 'key'
- [x] Add an endpoint to generate an api-key 
- [x] Add an enpoint to remove an api-key

- [x] In profile page add a field for viewing an api key
- [x] In profile page add button to generate an api key

### PR DESCRIPTION
**Pull request by sanmont3drepo**
This fixes #1339

### CODE REVIEW

File: frontend/modules/currentUser/currentUser.redux.ts [hunk 1]
```diff
@@ -31,8 +31,10 @@ export const { Types: CurrentUserTypes, Creators: CurrentUserActions } = createA
 	updateButtonText: ['value'],
 	uploadAvatar: ['file'],
 	refreshAvatar: ['avatarUrl'],
+	generateApiKey: [],
+	deleteApiKey: [],
 	setAsInitialized: []
-}, { prefix: 'TEAMSPACE_' });
+}, { prefix: 'CURRENT_USER_' });
```
Comment (simplypixi) [inline frontend/modules/currentUser/currentUser.redux.ts]: 👍

**Comment by simplypixi** [comment]
image

Why did you split `showErrorDialog` into two different actions? If you had any reason to do that, why didn't you updated already migrated code?
Moreover, we need to be informed about such changes - please add us to the PR next time, at least. ;)

File: frontend/routes/dashboard/dashboard.component.tsx [hunk 2]
```diff
@@ -115,7 +112,7 @@ export class Dashboard extends React.PureComponent<IProps, any> {
 				<Sidebar>
 					<UserInfo
 						{...currentUser}
-						loading={!isInitialized && (isPending || isAvatarPending)}
+						loading={false}
```
Comment (simplypixi) [inline frontend/routes/dashboard/dashboard.component.tsx]: You can remove this prop ;)

File: frontend/routes/dashboard/dashboard.component.tsx [hunk 3]
```diff
@@ -96,10 +96,7 @@ export class Dashboard extends React.PureComponent<IProps, any> {
 
 	public renderDashboardRoute = ({match}) => (
 		<Content>
-			{ !this.props.isInitialized && this.props.isPending
```
Comment (simplypixi) [inline frontend/routes/dashboard/dashboard.component.tsx]: Is it correct?

File: frontend/routes/profile/profile.styles.ts [hunk 4]
```diff
@@ -45,6 +45,28 @@ export const StyledButton = styled(Button)`
...
+	}
+
+	& > button {
+		margin-right: 10px;
+	}
+`;
+
+export const DeleteButton = styled(Button)`
+  && {
+    color: rgba(255, 255, 255, 0.87);
```
Comment (simplypixi) [inline frontend/routes/profile/profile.styles.ts]: `${COLORS.WHITE_87}`

File: frontend/routes/userManagement/userManagement.component.tsx [hunk 5]
```diff
@@ -141,6 +147,7 @@ export class UserManagement extends React.PureComponent<IProps, IState> {
 		const teamspacesChanged = !isEqual(this.props.teamspaces, prevProps.teamspaces);
 		if (teamspacesChanged) {
 			changes.teamspacesItems = this.props.teamspaces.map(({account}) => ({value: account}));
+			this.props.onTeamspaceChange( this.props.match.params.teamspace);
```
Comment (simplypixi) [inline frontend/routes/userManagement/userManagement.component.tsx]:
```suggestion
			this.props.onTeamspaceChange(match.params.teamspace);
```

**Comment by carmenfan** [comment]
> image
> 
> Why did you split `showErrorDialog` into two different actions? If you had any reason to do that, why didn't you updated already migrated code?
> Moreover, we need to be informed about such changes - please add us to the PR next time, at least. ;)

This was due to the past `showErrorDialog` assumed it was an endpoint error and always took parameters such as error codes which did not exist. And @sanmont3drepo had a need for reusing the error dialog for a UI level error to show to the client, thus he asked if he can change it and I thought it made sense. He did change all the other functions to use the Endpoint one, but I guess some newer functionalities you guys were working on wasn't merged in yet so the change was not applied.

Apologies, this happened in a PR over christmas so the conversations got lost (mainly me forgetting to mention it post xmas 😨) . We'll make sure we'll contact you about further changes like this !

File: frontend/routes/dashboard/dashboard.component.tsx (hunk 2 above)
Comment (sanmont3drepo) [reply frontend/routes/dashboard/dashboard.component.tsx]: That was for debugging that I forgot to revert.

File: frontend/routes/dashboard/dashboard.component.tsx (hunk 3 above)
Comment (sanmont3drepo) [reply frontend/routes/dashboard/dashboard.component.tsx]: Yes, we are moving from having most of the data in the /teamspace.json , to several endpoints with less payload. That means that each tab will have  the responsibility of knowing which part of the state it needs and load it accordingly, that also means the loading state for each tab varies.

**Comment by simplypixi** [comment]
> > image
> > Why did you split `showErrorDialog` into two different actions? If you had any reason to do that, why didn't you updated already migrated code?
> > Moreover, we need to be informed about such changes - please add us to the PR next time, at least. ;)
> 
> This was due to the past `showErrorDialog` assumed it was an endpoint error and always took parameters such as error codes which did not exist. And @sanmont3drepo had a need for reusing the error dialog for a UI level error to show to the client, thus he asked if he can change it and I thought it made sense. He did change all the other functions to use the Endpoint one, but I guess some newer functionalities you guys were working on wasn't merged in yet so the change was not applied.

Ok, got it ;)

> Apologies, this happened in a PR over christmas so the conversations got lost (mainly me forgetting to mention it post xmas 😨) . We'll make sure we'll contact you about further changes like this!

No problem! I just try to avoid situations when we'll have to fix our code because of some unexpected changes. ;)

File: frontend/routes/dashboard/dashboard.component.tsx (hunk 3 above)
Comment (simplypixi) [reply frontend/routes/dashboard/dashboard.component.tsx]: Awesome!

**Pull request merged**